README.md
# status-servlet [![Build Status](https://travis-ci.org/UnboundID/status-servlet.svg?branch=master)](https://travis-ci.org/UnboundID/status-servlet)

This is a status servlet extension for the Data Governance Server. It may be
used as the health check target for a layer 7 HTTP load balancer such as HAProxy
or Amazon Elastic Load Balancer and can generally be considered a more reliable
indicator of server availability than an arbitrary service path.

## Usage

Request the servlet at the configured path. 
For example, `https://server/status`. If the Data Governance Server's services
are available, a 200 OK will be returned. Otherwise, a 503 SERVICE UNAVAILABLE
will be returned.

```http
GET /status HTTP/1.1
Accept: application/json
Content-Type: application/json; charset=utf-8
Host: example.com



HTTP/1.1 200 OK
Content-Type: application/json
Date: Sun, 05 Jun 2016 03:59:14 GMT
Transfer-Encoding: chunked

{
    "loadBalancingAlgorithms": [
        {
            "available": true, 
            "name": "User Store LBA", 
            "numAvailableServers": 2, 
            "numDegradedServers": 0, 
            "numUnavailableServers": 0
        }
    ], 
    "servlets": [
        {
            "enabled": true, 
            "name": "SCIM2"
        }, 
        {
            "enabled": true, 
            "name": "UserInfo"
        }, 
        {
            "enabled": true, 
            "name": "JWK"
        }, 
        {
            "enabled": true, 
            "name": "Authentication"
        }, 
        {
            "enabled": true, 
            "name": "OAuth2"
        }
    ], 
    "storeAdapters": [
        {
            "available": true, 
            "name": "UserStoreAdapter"
        }
    ]
}
```

## Installation

Java 7 or up is required.

First, build the extension bundle by running `mvn package`. The extension bundle will be saved as a zip file in the `target` directory.

Next, install the extension bundle.

```
manage-extension --install com.unboundid.ops.status-servlet-VERSION.zip
```

Then configure the extension, assign it to an HTTPS connection handler and 
restart the connection handler.

```
dsconfig create-http-servlet-extension --extension-name Status \
  --type third-party --set "description:Reports Data Governance Server service availability" \
  --set extension-class:com.unboundid.ops.StatusServletExtension \
  --set extension-argument:path=/status \
  --set extension-argument:monitored-servlet=OAuth2 \
  --set extension-argument:monitored-servlet=SCIM2 \
  --set extension-argument:monitored-servlet=UserInfo \
  --set extension-argument:monitored-servlet=JWK \
  --set extension-argument:monitored-servlet=Authentication
dsconfig set-connection-handler-prop --handler-name "HTTPS Connection Handler" \
  --set enabled:false --add http-servlet-extension:Status
dsconfig set-connection-handler-prop --handler-name "HTTPS Connection Handler" \
  --set enabled:true
```

The `create-http-servlet-extension` command above takes the following `extension-argument` values:

| Extension argument | Description |
| --- | --- |
| path | The servlet's path. Defaults to `/status`. |
| monitored-servlet | The name of a servlet to monitor. This corresponds to an entry in the Broker's _HTTP Servlet Extensions_ configuration. If the servlet name ends with the word 'Servlet', this should be omitted. For example, 'OAuth2 Servlet' is specified as 'OAuth2'. This argument may be specified multiple times. If not specified, then no servlets are monitored. |

## Support and reporting bugs

This is an unsupported example, but support will be provided on a best-effort basis.

Please report issues using the project's [issue tracker](https://github.com/UnboundID/status-servlet/issues).

## License

This is licensed under the Apache License 2.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [Unreleased]
### Changed
- Updated connectionFailureTest() to execute more reliably.
- Updated generated documentation assembly to use Ping Identity branding.

## [1.1.1] - 2016-10-04
### Changed
- Initial public release.

src/main/java/com/unboundid/ops/StatusServletExtension.java
/*
 * Copyright 2016 UnboundID Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.unboundid.ops;

import com.unboundid.directory.sdk.http.api.HTTPServletExtension;
import com.unboundid.directory.sdk.http.config.HTTPServletExtensionConfig;
import com.unboundid.directory.sdk.http.types.HTTPServerContext;
import com.unboundid.ldap.sdk.LDAPException;
import com.unboundid.util.args.ArgumentException;
import com.unboundid.util.args.ArgumentParser;
import com.unboundid.util.args.StringArgument;

import javax.servlet.http.HttpServlet;
import java.util.Collections;
import java.util.List;

/**
 * An {@link HTTPServletExtension} for the {@link StatusServlet}, which
 * reports a server's status, and can be used as a health check endpoint for
 * HTTP load balancers.
 *
 * @author Jacob Childress
 */
public class StatusServletExtension extends HTTPServletExtension
{
  private static final String ARG_PATH = "path";
  private static final String ARG_MONITORED_SERVLET = "monitored-servlet";

  private String path;


  /** {@inheritDoc} */
  @Override
  public String getExtensionName()
  {
    return "Status Servlet";
  }


  /** {@inheritDoc} */
  @Override
  public String[] getExtensionDescription()
  {
    return new String[] {
            "This extension provides a status servlet. It reports status for " +
                "store adapters, LDAP store adapters, and a configurable set of " +
                "HTTP servlets. The status servlet (default path '/status') can " +
                "be used as a health check target for HTTP load balancers. The " +
                "servlet will return a 200 OK if the server's services are " +
                "available and a 503 SERVICE UNAVAILABLE if they are not."
    };
  }


  /** {@inheritDoc} */
  @Override
  public void defineConfigArguments(final ArgumentParser parser)
          throws ArgumentException
  {
    parser.addArgument(new StringArgument(
            null, ARG_PATH, false, 1, "{path}",
            "The base path of this servlet. The default is '/status'.",
            "/status"));
    parser.addArgument(new StringArgument(
            null, ARG_MONITORED_SERVLET, false, 0, "{servletName}",
            "The name of a servlet that is expected to be enabled. " +
                    "By default, no servlets are monitored."));
  }


  /** {@inheritDoc} */
  @Override
  public HttpServlet createServlet(
          HTTPServerContext httpServerContext,
          HTTPServletExtensionConfig httpServletExtensionConfig,
          ArgumentParser argumentParser) throws LDAPException
  {
    StringArgument pathArgument =
            (StringArgument) argumentParser.getNamedArgument(ARG_PATH);
    path = pathArgument.getValue();
    StringArgument monitoredServlets =
            (StringArgument) argumentParser.getNamedArgument(ARG_MONITORED_SERVLET);
    return new StatusServlet(httpServerContext,
                             httpServerContext.getInternalRootConnection(),
                             monitoredServlets.getValues().toArray(
                                           new String[monitoredServlets.getValues().size()]));
  }


  /** {@inheritDoc} */
  @Override
  public List<String> getServletPaths()
  {
    return Collections.singletonList(path);
  }
}
